pkg/registrymirror/registrymirrorsstore.go

```diff
@@ -0,0 +1,325 @@
+
+// WithConfigPath sets the path to read/write the consolidated registries config.
+func WithConfigPath(path string) fileStoreOption {
+	return func(s *FileStore) { s.configPath = path }
+}
+
+// WithDelay sets a delay before the updated config is written to disk.
+func WithDelay(delay time.Duration) fileStoreOption {
+	return func(s *FileStore) { s.updateDelay = delay }
+}
+
+// NewFileStore creates a new FileStore.
+func NewFileStore(opts ...fileStoreOption) *FileStore {
+	s := &FileStore{
+		configPath:  defaultRegistriesPath,
+		updateDelay: defaultDelay,
+		icspRules:   make(map[types.UID]*operatorV1Alpha1.ImageContentSourcePolicy),
+		idmsRules:   make(map[types.UID]*configV1.ImageDigestMirrorSet),
+		itmsRules:   make(map[types.UID]*configV1.ImageTagMirrorSet),
+	}
+
+	for _, opt := range opts {
+		opt(s)
+	}
+
+	s.systemContext = &ciTypes.SystemContext{SystemRegistriesConfPath: s.configPath}
+	return s
+}
+
+// Cleanup resets the store which includes in-memory and disk resources.
+func (s *FileStore) Cleanup() {
+	s.ruleRWMutex.Lock()
+	defer s.ruleRWMutex.Unlock()
+
+	s.icspRules = make(map[types.UID]*operatorV1Alpha1.ImageContentSourcePolicy)
+	s.idmsRules = make(map[types.UID]*configV1.ImageDigestMirrorSet)
+	s.itmsRules = make(map[types.UID]*configV1.ImageTagMirrorSet)
+
+	if err := os.Remove(s.configPath); err != nil && !os.IsNotExist(err) {
+		log.Warnf("Failed to cleanup registries config at %q: %v", s.configPath, err)
+	}
+	sysregistriesv2.InvalidateCache()
+}
+
+func (s *FileStore) updateConfig() error {
+	if s.updateDelay == 0 {
+		return s.updateConfigNow()
+	}
+
+	return s.updateConfigDelayed()
+}
+
+func (s *FileStore) updateConfigDelayed() error {
+	// Using a different mutex to protect timer vs. rules to avoid deadlock.
+	s.timerMutex.Lock()
+	defer s.timerMutex.Unlock()
+
+	if s.timer != nil {
+		s.timer.Stop()
+	}
+
+	s.timer = time.AfterFunc(s.updateDelay, func() {
+		if err := s.updateConfigNow(); err != nil {
+			log.Errorf("Failed to update the registry mirror config: %v", err)
+		}
+	})
+
+	return nil
+}
+
+func (s *FileStore) updateConfigNow() error {
+	icspRules, idmsRules, itmsRules := s.getAllMirrorSets()
+
+	// Populate config.
+	config := new(sysregistriesv2.V2RegistriesConf)
+	err := registries.EditRegistriesConfig(config, nil, nil, icspRules, idmsRules, itmsRules)
+	if err != nil {
+		return errors.Wrap(err, "could not create registries config")
+	}
+
+	// Encode config.
+	var newData bytes.Buffer
+	encoder := toml.NewEncoder(&newData)
+	if err := encoder.Encode(config); err != nil {
+		return errors.Wrap(err, "could not encode registries config")
+	}
+
+	// Ensure output dir exists.
+	err = os.MkdirAll(filepath.Dir(s.configPath), 0755)
+	if err != nil {
+		return errors.Wrap(err, "could not create directories")
+	}
+
+	// Open handle to output file.
+	f, err := os.Create(s.configPath)
+	if err != nil {
+		return errors.Wrap(err, "could not create/open file")
+	}
+	defer utils.IgnoreError(f.Close)
```

Comment:
this is a no-op if this is the second time, right?

Reply:
Unclear from the source if its a no-op, is not causing issues and is the preferred way to handle file closes per the ACS style guide.

Reply:
Stepped through with a debugger and IMO it can be considered a no-op.  The first call will perform the actual close (and assumed flush).  The deferred call will produce `ErrClosed` indicating the file is already closed - so no harm done.